packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * @typedef {Object} CopyError
+ * Corresponding PassStyle is "copyError".
+ *
+ * TODO What I'd like to say in these types in that only a CopyError
+ * without an `errorId` field is a Comparable.
  *
- * An OnlyData value is PureData when it contains no hidden mutable state,
- * e.g., when none of its pass-by-copy composite data objects are proxies. This
- * cannot be determined by inspection. It can only be achieved by trusted
- * construction. A PureData value cannot be used as a communications channel,
- * and can therefore be safely shared with subgraphs that should not be able
- * to communicate with each other.
+ * ***TODO WARNING*** Since CopyErrors have no analog amount Records and
+ * Tuples, perhaps a CopyError in general should only be a Comparable but
+ * not a PassableKey. Or perhaps not even a Comparable! In that case we
+ * wouldn't need to make a special case for `errorId`.
+ * @property {string} name
+ * @property {string} message
+ * @property {string=} errorId
  */
 
 /**
- * @typedef {*} Remotable
- * Might be an object explicitly deemed to be `Remotable`, an object inferred
- * to be Remotable, or a remote presence of a Remotable.
+ * @template {Passable} T
+ * @typedef { LeafData | CopyArray<T> | CopyRecord<T> | CopyError
+ * } NestedData
+ * Corresponding PassStyle values are
+ * the PassStyles of LeafData, "copyArray", "copyRecord", "copyError"
+ *
+ * NestedData is parameterized over various "deep" constraints. See
+ * the discussion in CopyArray and CopyRecord above.
  */
 
 /**
- * @typedef {Promise | Remotable} PassableCap
- * The leaves of a Passable's pass-by-copy superstructure.
+ * @typedef {NestedData<PureData>} PureData
+ * Contains no capabilities (promises, remotables)
+ * Contains no sets or maps
+ * Contains no proxies, or anything which can sense being examined.
+ * Cannot be tested for, but only assured by construction
+ * pureCopy(OnlyData) => PureData
+ */
+
+/**
+ * @typedef {PureData | NestedData<OnlyData>} OnlyData
+ * Tested by `isOnlyData`, `assertOnlyData`
+ * Contains no capabilities (promises, remotables)
+ * Contains no sets or maps
+ * The objects it contains may contain proxies, and therefore
+ * may cause side effects when examined.
+ */
+
+/**
+ * @typedef {Remotable | OnlyData | NestedData<PassableKey> } PassableKey
+ *
+ * PassableKeys can be compared for equivalence with `sameKey`. A
+ * PassableKey can be used as a key in a StrongStore, i.e., used
+ * as a key in a PassableMap, MemoryMap, or VirtualMap, and used as
+ * an element of a PassableSet, MemorySet, or VirtualSet.
+ *
+ * Note that a PassableKey cannot contain promises nor passable
+ * stores.
+ *
+ * Passing should preserve `sameKey` equivalence. Given that PassableKeys
+ * `Xa` and `Ya` in vat A, when passed to vat B, arrive as `Xb` and `Yb`, then
+ * `sameKey(Xa, Ya)` in vat A must have the same answer as `sameKey(Xb, Yb)`
+ * in vat B.
+ */
+
+/**
+ * @template {PassableKey} K
+ * @typedef {*} PassableStore
```

Comment:
The convention for objects with methods is:
```suggestion
 * @typedef {Object} PassableStore
```

In general, the `*` type is usually reserved for "I don't know what type this should be", but `any` means "I mean for this really to be anything" and `Object` is "I mean for this to be any kind of typeof 'object'", but `Record<K,V>` means a more specific object, and `{a:B,c:D}` means even more specificity.

It would be nice to document this somewhere.  I guess I just did. 🙂